internal/topic/topicreaderinternal/stream_reader_impl.go

```diff
@@ -541,29 +543,40 @@ func (r *topicStreamReaderImpl) dataRequestLoop(ctx context.Context) {
 			_ = r.CloseWithError(ctx, r.ctx.Err())
 			return
 
-		case free := <-r.freeBytes:
-			sum := free
-
-			// consume all messages from order and compress it to one data request
-		forConsumeRequests:
-			for {
-				select {
-				case free = <-r.freeBytes:
-					sum += free
-				default:
-					break forConsumeRequests
-				}
+		case <-r.freeBytes:
+			freeSpacePercentage, err := r.getFreeSpacePercentage()
+			if err != nil {
+				return
 			}
 
-			resCapacity := r.addRestBufferBytes(sum)
-			trace.TopicOnReaderSentDataRequest(r.cfg.Trace, r.readConnectionID, sum, resCapacity)
-			if err := r.sendDataRequest(sum); err != nil {
+			if freeSpacePercentage <= FreeSpacePercentageThreshold {
+				return
+			}
+
+			freeSpaceBytes := r.cfg.BufferSizeProtoBytes - r.getRestBufferBytes()
+			resCapacity := r.addRestBufferBytes(freeSpaceBytes)
+			trace.TopicOnReaderSentDataRequest(r.cfg.Trace, r.readConnectionID, freeSpaceBytes, resCapacity)
+
+			if err := r.sendDataRequest(freeSpaceBytes); err != nil {
 				return
 			}
 		}
 	}
 }
 
+var errCannotCalcFreeSpacePercentage = errors.New("cannot calculate free space percentage")
+
+func (r *topicStreamReaderImpl) getFreeSpacePercentage() (percentage int, err error) {
+	sizeBytes := float64(r.cfg.BufferSizeProtoBytes)
+	usedSpaceBytes := float64(r.getRestBufferBytes())
```

Comment:
getRestBufferBytes return free bytes in buffer, not used.


what about rename the function?

and the value is bad for request data, because in show real buffer usage (decrease when receive data from server and increase when return data from sdk).
It will race conditions and request a lot of data over buffer size.

need another counter for reflect bytes, which sdk can request from server. It is not same as real usage.

the getRestBufferBytes usable for monitoring purposes only.